core/src/main/scala/org/apache/spark/scheduler/TaskAssigner.scala

```diff
@@ -0,0 +1,232 @@
+    offer = Random.shuffle(workOffer.map(o => new OfferState(o)))
+  }
+
+  /** Invoked at each round of Taskset assignment to initialize the internal structure. */
+  def init(): Unit
+
+  /**
+   * Tests whether there is offer available to be used inside of one round of Taskset assignment.
+   * @return  `true` if a subsequent call to `next` will yield an element,
+   *          `false` otherwise.
+   */
+  def hasNext: Boolean
+
+  /**
+   * Produces next worker offer based on the task assignment strategy.
+   * @return  the next available offer, if `hasNext` is `true`,
+   *          undefined behavior otherwise.
+   */
+  def next(): OfferState
+
+  /**
+   * Invoked by the TaskScheduler to indicate whether the current offer is accepted or not so that
+   * the assigner can decide whether the current worker is valid for the next offering.
+   *
+   * @param isAccepted whether TaskScheduler assigns a task to current offer.
+   */
+  def offerAccepted(isAccepted: Boolean): Unit
+}
+
+object TaskAssigner extends Logging {
+  private val roundrobin = classOf[RoundRobinAssigner].getCanonicalName
+  private val packed = classOf[PackedAssigner].getCanonicalName
+  private val balanced = classOf[BalancedAssigner].getCanonicalName
+  private val assignerMap: Map[String, String] =
+    Map("roundrobin" -> roundrobin,
+      "packed" -> packed,
+      "balanced" -> balanced)
+
+  def init(conf: SparkConf): TaskAssigner = {
+    val assignerName = conf.get(config.SPARK_SCHEDULER_TASK_ASSIGNER.key, "roundrobin")
+    val className = {
+      val name = assignerMap.get(assignerName.toLowerCase())
+      name.getOrElse {
+        logWarning(s"$assignerName cannot be constructed, fallback to default $roundrobin.")
+        roundrobin
+      }
+    }
+    // The className is valid. No need to catch exceptions.
+    logInfo(s"Constructing TaskAssigner as $className")
+    Utils.classForName(className).getConstructor().newInstance().asInstanceOf[TaskAssigner]
+      .withCpuPerTask(cpuPerTask = conf.getInt("spark.task.cpus", 1))
+  }
+}
+
+/**
+ * Assigns the task to workers with available cores in a roundrobin manner.
+ */
+class RoundRobinAssigner extends TaskAssigner {
+  private var currentOfferIndex = 0
+
+  override def init(): Unit = {
+    currentOfferIndex = 0
+  }
+
+  override def hasNext: Boolean = currentOfferIndex < offer.size
+
+  override def next(): OfferState = {
+    offer(currentOfferIndex)
+  }
+
+  override def offerAccepted(isAccepted: Boolean): Unit = {
+    currentOfferIndex += 1
+  }
+}
+
+/**
+ * Assigns the task to workers with the most available cores. In other words, BalancedAssigner tries
+ * to distribute the task across workers in a balanced way. Potentially, it may alleviate the
+ * workers' memory pressure as less tasks running on the same workers, which also indicates that
+ * the task itself can make use of more computation resources, e.g., hyper-thread, across clusters.
+ */
+class BalancedAssigner extends TaskAssigner {
```

Comment:
@rxin These two do the opposite thing. The packed scheduler tries to schedule tasks to workers as few as possible so that some workers without task running can be released. The balanced assigner tries to schedule the tasks to workers with least work load.

If user wants optimal resource reservation, they may want the packer assigner. If user want observe some memory pressure, they may want to try the balanced assigner.